Makefile

```diff
@@ -1,28 +1,59 @@
-ifndef GOPATH
-$(error No GOPATH set)
-endif
+SHORT_NAME := helm
+DEIS_REGISTRY ?= ${DEV_REGISTRY}
+IMAGE_PREFIX ?= helm
 
+REPO_PATH := github.com/helm/${SHORT_NAME}
+
+# The following variables describe the containerized development environment
+# and other build options
 BIN_DIR := bin
 DIST_DIR := _dist
 GO_PACKAGES := action chart config dependency log manifest release plugins/sec plugins/example codec
 MAIN_GO := helm.go
 HELM_BIN := $(BIN_DIR)/helm
 PATH_WITH_HELM = PATH="$(shell pwd)/$(BIN_DIR):$(PATH)"
-
 VERSION := $(shell git describe --tags --abbrev=0 2>/dev/null)+$(shell git rev-parse --short HEAD)
+DEV_ENV_IMAGE := quay.io/deis/go-dev:0.9.0
+DEV_ENV_WORK_DIR := /go/src/${REPO_PATH}
+DEV_ENV_CMD := docker run --rm -v ${CURDIR}:${DEV_ENV_WORK_DIR} -w ${DEV_ENV_WORK_DIR} ${DEV_ENV_IMAGE}
+DEV_ENV_CMD_INT := docker run -it --rm -v ${CURDIR}:${DEV_ENV_WORK_DIR} -w ${DEV_ENV_WORK_DIR} ${DEV_ENV_IMAGE}
+LDFLAGS := "-s -X main.version=${VERSION}"
+
+
+
+# Allow developers to step into the containerized development environment
+
+# Containerized dependency resolution
+dev: check-docker
+	${DEV_ENV_CMD_INT} bash
+
+dev-registry: check-docker
+	@docker inspect registry >/dev/null 2>&1 && docker start registry || docker run --restart="always" -d -p 5000:5000 --name registry registry:0.9.1
+	@echo
+	@echo "To use a local registry for Deis development:"
+	@echo "    export DEIS_REGISTRY=`docker-machine ip $$(docker-machine active 2>/dev/null) 2>/dev/null || echo $(HOST_IPADDR) `:5000/"
+
+bootstrap: 
+	${DEV_ENV_CMD} glide install
 
-export GO15VENDOREXPERIMENT=1
+# Containerized build of the binary
+build: check-docker
+	mkdir -p ${BIN_DIR}
+	${DEV_ENV_CMD} make binary-build
 
-ifndef VERSION
-  VERSION := git-$(shell git rev-parse --short HEAD)
-endif
+docker-build: build check-docker
+	docker build --rm -t ${IMAGE} rootfs
+	docker tag -f ${IMAGE} ${MUTABLE_IMAGE}
 
-build: $(MAIN_GO)
-	go build -o $(HELM_BIN) -ldflags "-X github.com/helm/helm/cli.version=${VERSION}" $<
 
-bootstrap:
-	go get -u github.com/golang/lint/golint github.com/mitchellh/gox
-	glide install
+# Builds the binary-- this should only be executed within the
+# containerized development environment.
+
+binary-build:
+	make build-all
+	@cd $(DIST_DIR) && \
+	find * -type d -exec zip -jr helm-$(VERSION)-{}.zip {} \; && \
+	cd -
 
 build-all:
 	gox -verbose \
```

Comment:
@chaitanyaenr don't forget, this still needs to be done inside a container. For example, `${DEV_ENV_CMD} gox ...`